addons/portal/controllers/portal.py

```diff
@@ -155,55 +154,116 @@ def counters(self, counters, **kw):
+        elif action == "read":
+            data = {key: getattr(contact, key) for key in fields}
+        elif action == "delete":
+            contact.unlink()
+            return request.redirect('/my/account')
+        if data:
+            error = error_message = None
+            error, error_message = self.details_form_validate(data, contact and contact.id == partner.id)
             values.update({'error': error, 'error_message': error_message})
-            values.update(post)
-            if not error:
-                values = {key: post[key] for key in self.MANDATORY_BILLING_FIELDS}
-                values.update({key: post[key] for key in self.OPTIONAL_BILLING_FIELDS if key in post})
-                for field in set(['country_id', 'state_id']) & set(values.keys()):
-                    try:
-                        values[field] = int(values[field])
-                    except:
-                        values[field] = False
-                values.update({'zip': values.pop('zipcode', '')})
-                partner.sudo().write(values)
+            for field in set(['country_id', 'state_id']):
+                try:
+                    data[field] = int(data[field])
+                except ValueError:
+                    data[field] = False
+            if action == "edit" and not error:
+                if not partner_id:
+                    data['parent_id'] = partner.id
+                    Partner.create(data)
+                else:
+                    contact.sudo().write(data)
                 if redirect:
                     return request.redirect(redirect)
-                return request.redirect('/my/home')
+                return request.redirect('/my/account')
+            values.update(data)
+            values['zipcode'] = values.pop('zip', False)
 
         countries = request.env['res.country'].sudo().search([])
         states = request.env['res.country.state'].sudo().search([])
 
         values.update({
             'partner': partner,
+            'contact': contact,
             'countries': countries,
             'states': states,
             'has_check_vat': hasattr(request.env['res.partner'], 'check_vat'),
             'redirect': redirect,
             'page_name': 'my_details',
         })
 
-        response = request.render("portal.portal_my_details", values)
+        response = request.render("portal.portal_my_address", values)
         response.headers['X-Frame-Options'] = 'DENY'
         return response
 
+    @route('/my/users_accesses', type='http', auth='user', website=True, methods=['GET', 'POST'])
+    def user_accesses(self, **kw):
+        values = self._prepare_portal_layout_values()
+        values['page_name'] = 'users_accesses'
+        partner = request.env.user.partner_id
+        partners = partner.commercial_partner_id.child_ids
+        portal_wizard = request.env['portal.wizard'].sudo().create({'partner_ids': partners})
```

Comment:
Not sure manually creating wizards is a common practice, and that it will work as intended.

Maybe we should add a separate commit that moves the "action_grant_access" method from the wizard to the partner model to be able to call it easily?

Because again, this involves a lot of sudo, and I think this route could be abused to grant portal access to an arbitrary email address that matches a partner.